### PR TITLE
QMCHamiltonian with Listener support

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

API changes and additions to QMCHamiltonian to support per particle Listerners from Estimators.
Next PR will add an estimator that logs per particle values to illustrate the API's use.

## What type(s) of changes does this code introduce?

- New feature
- Documentation changes

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by PDoakORNL** [comment]
test this please

**Comment by PDoakORNL** [comment]
test this please

File: src/QMCHamiltonians/Listener.hpp [hunk 1]
```diff
@@ -69,15 +69,13 @@ class ListenerVector
...
-    ListenerOption(const std::vector<ListenerVector<T>>& le, const std::vector<ListenerVector<T>>& li)
-        : electrons(le), ions(li)
-    {}
-    const std::vector<ListenerVector<T>>& electrons;
-    const std::vector<ListenerVector<T>>& ions;
+  ListenerOption(const std::vector<ListenerVector<T>>& le, const std::vector<ListenerVector<T>>& li)
+      : electrons(le), ions(li)
+  {}
+  const std::vector<ListenerVector<T>>& electrons;
+  const std::vector<ListenerVector<T>>& ions;
```
Comment (ye-luo) [inline src/QMCHamiltonians/Listener.hpp:76]: electrons, ions are confusing names.

File: src/QMCHamiltonians/QMCHamiltonian.cpp [hunk 2]
```diff
@@ -595,10 +651,9 @@ std::vector<QMCHamiltonian::FullPrecRealType> QMCHamiltonian::mw_evaluate(
   //   pset.PropertyList[WP::LOCALENERGY]    = ham.LocalEnergy;
   //   pset.PropertyList[LOCALPOTENTIAL] = ham.LocalEnergy - ham.KineticEnergy;
   // };
-
+  const auto HC_list(extract_HC_list(ham_list, 0));
   for (int iw = 0; iw < ham_list.size(); iw++)
   {
-    const auto HC_list(extract_HC_list(ham_list, 0));
     updateKinetic(HC_list[iw], ham_list[iw], p_list[iw]);
   }
```
Comment (ye-luo) [inline src/QMCHamiltonians/QMCHamiltonian.cpp:658]: Remove brackets?

File: src/QMCHamiltonians/QMCHamiltonian.cpp [hunk 3]
```diff
@@ -789,12 +844,31 @@ std::vector<QMCHamiltonian::FullPrecRealType> QMCHamiltonian::mw_evaluateWithTop
 
   auto& ham_leader            = ham_list.getLeader();
   const int num_ham_operators = ham_leader.H.size();
-  for (int i_ham_op = 0; i_ham_op < num_ham_operators; ++i_ham_op)
+
+  int kinetic_index = 0;
```
Comment (ye-luo) [inline src/QMCHamiltonians/QMCHamiltonian.cpp:848]: Add const?

File: src/QMCHamiltonians/QMCHamiltonian.cpp [hunk 4]
```diff
@@ -985,6 +1059,8 @@ std::vector<int> QMCHamiltonian::mw_makeNonLocalMoves(const RefVectorWithLeader<
 
 void QMCHamiltonian::createResource(ResourceCollection& collection) const
 {
+  auto new_res        = std::make_unique<QMCHamiltonianMultiWalkerResource>();
+  auto resource_index = collection.addResource(std::move(new_res));
```
Comment (ye-luo) [inline src/QMCHamiltonians/QMCHamiltonian.cpp:1063]: Merge the above two lines?

File: src/QMCHamiltonians/QMCHamiltonian.cpp [hunk 5]
```diff
@@ -1004,14 +1084,15 @@ void QMCHamiltonian::releaseResource(ResourceCollection& collection,
...
+  collection.takebackResource(std::move(ham_leader.mw_res_));
   for (int i_ham_op = 0; i_ham_op < ham_leader.H.size(); ++i_ham_op)
   {
     const auto HC_list(extract_HC_list(ham_list, i_ham_op));
     ham_leader.H[i_ham_op]->releaseResource(collection, HC_list);
   }
 }
 
-std::unique_ptr<QMCHamiltonian> QMCHamiltonian::makeClone(ParticleSet& qp, TrialWaveFunction& psi)
+std::unique_ptr<QMCHamiltonian> QMCHamiltonian::makeClone(ParticleSet& qp, TrialWaveFunction& psi) const
```
Comment (ye-luo) [inline src/QMCHamiltonians/QMCHamiltonian.cpp:1095]: Like this added security.

File: src/QMCHamiltonians/tests/test_Listener.cpp [hunk 6]
```diff
@@ -42,7 +46,7 @@ class Talker
   }
 };
 
-class TestReceiver
+class MockPerParticleEstimator
```
Comment (ye-luo) [inline src/QMCHamiltonians/tests/test_Listener.cpp:49]: Thanks for updating all these names.

File: src/QMCHamiltonians/tests/test_QMCHamiltonian.cpp [hunk 7]
```diff
@@ -42,6 +53,224 @@ TEST_CASE("QMCHamiltonian::flex_evaluate", "[hamiltonian]")
...
+}
+
+#ifndef QMC_CUDA
+
+/** QMCHamiltonian + Hamiltonians with listeners integration test
+ */
+TEST_CASE("integrateListeners", "[hamiltonian]")
+{
+  Communicate* comm;
+  comm = OHMMS::Controller;
```
Comment (ye-luo) [inline src/QMCHamiltonians/tests/test_QMCHamiltonian.cpp:66]: write above two in one line.

File: src/QMCHamiltonians/tests/test_QMCHamiltonian.cpp [hunk 8]
```diff
@@ -42,6 +53,224 @@ TEST_CASE("QMCHamiltonian::flex_evaluate", "[hamiltonian]")
...
+    }
+
+    // When only EE and EI coulomb potentials the local energy is just the sum of
+    // the local_pots and kinetic
+    auto sum_local_pots = std::accumulate(local_pots.begin(), local_pots.end(), 0.0);
+    auto sum_kinetic    = std::accumulate(kinetic.begin(), kinetic.end(), 0.0);
+    auto sum_local_nrg  = std::accumulate(local_nrg.begin(), local_nrg.end(), 0.0);
+    CHECK(sum_local_nrg == Approx(sum_local_pots + sum_kinetic));
+  }
+  outputManager.resume();
```
Comment (ye-luo) [inline src/QMCHamiltonians/tests/test_QMCHamiltonian.cpp:272]: Remove this resume and the pause above. We should have suppressed the output by default in unit tests.

**Review by ye-luo — COMMENTED** [review]
Minor changes requested.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**